lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -192,6 +224,63 @@ constFoldFIRRTLBinaryOp(Operation *op, ArrayRef<Attribute> operands,
   return getIntAttr(resultType, resultValue);
 }
 
+/// Applies the canonicalization function `canonicalize` to the given operation.
+///
+/// Determines which (if any) of the operation's operands are constants, and
+/// provides them as arguments to the callback function. Any `invalidvalue` in
+/// the input is mapped to a constant zero. The value returned from the callback
+/// is used as the replacement for `op`, and an additional pad operation is
+/// inserted if necessary. Does nothing if the result of `op` is of unknown
+/// width, in which case the necessity of a pad cannot be determined.
+static LogicalResult canonicalizePrimOp(
+    Operation *op, PatternRewriter &rewriter,
+    const function_ref<OpFoldResult(ArrayRef<Attribute>)> &canonicalize) {
+  // Can only operate on FIRRTL primary operations.
+  if (op->getNumResults() != 1)
+    return failure();
+  auto type = op->getResult(0).getType().dyn_cast<FIRRTLType>();
+  if (!type)
+    return failure();
+
+  // Can only operate on operations with a known result width.
+  auto width = type.getBitWidthOrSentinel();
+  if (width < 0)
+    return failure();
+
+  // Determine which of the operands are constants.
+  SmallVector<Attribute, 3> constOperands;
+  constOperands.reserve(op->getNumOperands());
+  for (auto operand : op->getOperands()) {
+    Attribute attr;
+    if (auto *defOp = operand.getDefiningOp())
+      TypeSwitch<Operation *>(defOp)
+          .Case<ConstantOp, SpecialConstantOp, InvalidValueOp>(
+              [&](auto op) { attr = op.fold({}).template get<Attribute>(); });
```

Comment:
MSVC may not like the template inside a lambda. You may want to do a special run with the MSVC action using this branch to make sure it compiles. If this is a problem, I will usually work around it by removing the `auto`. In this case it's necessary as there's no parent type that can be used here. Worst case, this could use three cases with explicit types instead of `auto`.

Reply:
Good idea, will do a run!